src/System.Management.Automation/CoreCLR/CorePsAssemblyLoadContext.cs

```diff
@@ -240,7 +240,28 @@ internal static IntPtr NativeDllHandler(Assembly assembly, string libraryName)
         {
             s_nativeDllSubFolder ??= GetNativeDllSubFolderName(out s_nativeDllExtension);
             string folder = Path.GetDirectoryName(assembly.Location);
-            string fullName = Path.Combine(folder, s_nativeDllSubFolder, libraryName) + s_nativeDllExtension;
+
+            // Issue #20740 shows that some developers specify `.dll` in their DllImport attribute even though
+            // they ship native libraries for Linux and macOS, but they aren't found because the extension is now
+            // treated as part of the filename. If the libraryName contains a known extension for wrong OS, we
+            // remove it and add the OS specific extension.
+            var extension = Path.GetExtension(libraryName).ToLowerInvariant();
+            string fullName;
+
+            if (string.IsNullOrEmpty(extension))
+            {
+                fullName = Path.Combine(folder, s_nativeDllSubFolder, libraryName) + s_nativeDllExtension;
+            }
+            else if ((extension == ".dll" && !Platform.IsWindows) ||
+                (extension == ".so" && !Platform.IsLinux) ||
+                (extension == ".dylib" && !Platform.IsMacOS))
```

Comment:
Maybe use `s_nativeDllExtension`?
```suggestion
            else if (extension != s_nativeDllExtension)
```
And maybe we could make the code more readable like:
```c#
libraryName = extension switch
{
   s when string.IsNullOrEmpty(s) => libraryName + s_nativeDllExtension,
   s when s != s_nativeDllExtension => Path.GetFileNameWithoutExtension(libraryName) + s_nativeDllExtension,
   _ => libraryName;
}
string fullName = Path.Combine(folder, s_nativeDllSubFolder, libraryName);
return NativeLibrary.TryLoad(fullName, out IntPtr pointer) ? pointer : IntPtr.Zero;
```